### DISCUSSION TITLE
GGML CPU ONLY VS GGML with GPU Acceleration - Also includes three GPTQ Backend comparisons - If your curious about my results take a look.

### DISCUSSION THREAD

**Discussion by viperwasp** [oobabooga/text-generation-webui#2674] [General]
I recently got GPU Acceleration working on Windows 10, RTX 4080, 12700k. (Without WSL)
Many thanks to the kind individuals who helped me get this working. 
I know some people ask for this kind of data. 

1. GPU Offloading Layers were set so that my Video Ram would be around 90%-93% in use, giving overhead for fluctuation and multitasking. I kept threads set to 8 across all these tests. 

2. There is an important thing to mention that capitalizes on the benefit of GPU Acceleration beyond just Tokens Per Second. I will mention that in the first result section. 

3. The prompt, and answer are not that relevant. Not even the overall generation time. This is mostly just to look at tokens per second. Your results will vary depending on GPU, CPU, settings etc. 

4. I performed the same prompt 4 times for variance etc. For some reason the first time is always slower. So keep that in mind. 

**Test 1**
**Wizard-Vicuna-30B-Uncensored.ggmlv3.q5_0 CPU With GPU Accelerate**
What is the capital of Canada. An the capital of USA. And China and Russia and the UK.

**Answer**
The capital of Canada is Ottawa, Ontario. 
The capital of the United States is Washington D.C., which stands for District of Columbia. 
The capital of China is Beijing. 
The capital of Russia is Moscow. 
And the capital of the UK (United Kingdom) is London.

Output generated in 37.35 seconds (1.79 tokens/s, 67 tokens, context 61, seed 1195351953) Initial Run
Output generated in 23.62 seconds (2.84 tokens/s, 67 tokens, context 61, seed 719084501)
Output generated in 24.05 seconds (2.79 tokens/s, 67 tokens, context 61, seed 742883115)
Output generated in 23.24 seconds (2.88 tokens/s, 67 tokens, context 61, seed 1076689416)

**Wizard-Vicuna-30B-Uncensored.ggmlv3.q5_0 CPU Only**
What is the capital of Canada. An the capital of USA. And China and Russia and the UK.

**Answer**
The capital of Canada is Ottawa, Ontario. 
The capital of the United States is Washington D.C., which stands for District of Columbia. 
The capital of China is Beijing. 
The capital of Russia is Moscow. 
And the capital of the UK (United Kingdom) is London

Output generated in 57.45 seconds (1.17 tokens/s, 67 tokens, context 61, seed 983449324) initial Run
Output generated in 38.01 seconds (1.76 tokens/s, 67 tokens, context 61, seed 1344744116)
Output generated in 36.84 seconds (1.82 tokens/s, 67 tokens, context 61, seed 87118797)
Output generated in 37.20 seconds (1.80 tokens/s, 67 tokens, context 61, seed 1595453416)

**Results**
If I am not mistaken that is about a 58% increase give or take.
However I need to mention that at least for 30B models. When using **GGML CPU only**. I notice that the "Ooba" boggles down and basically stops working in longer chats. Likely related to context size/length of discussion. But it gets to the point were the system does not even start generating new tokens for 3-5 minutes. With GGML GPU Acceleration this is basically gone. I think the longest pauses even with context size 1800 tokens were about 20 seconds, even when submitting another prompt into a long conversation. This is day and night in comparison and for the first time ever I can fully utilize 30B models. So I am officially a 30 Billion man now. lol

--------------------------------------------------------
**Test 2**
**VicUnlocked-Alpaca-65B.ggmlv3.q4_0 CPU With GPU Accelerate**

What is the capital of Canada. An the capital of USA. And China and Russia and the UK.

**Answer**
The capital of Canada is Ottawa, Ontario. The capital of the United States is Washington D.C. The capital of China is Beijing. The capital of Russia is Moscow. The capital of the United Kingdom is London.

Output generated in 50.22 seconds (0.98 tokens/s, 49 tokens, context 58, seed 821096343) Initial Run
Output generated in 34.91 seconds (1.40 tokens/s, 49 tokens, context 58, seed 1137454528)
Output generated in 35.26 seconds (1.39 tokens/s, 49 tokens, context 58, seed 270104423)
Output generated in 34.53 seconds (1.42 tokens/s, 49 tokens, context 58, seed 508551429)

**VicUnlocked-Alpaca-65B.ggmlv3.q4_0 CPU ONLY** 

What is the capital of Canada. An the capital of USA. And China and Russia and the UK.

**Answer**
The capital of Canada is Ottawa, Ontario. The capital of the United States is Washington D.C., which stands for District of Columbia. The capital of China is Beijing, while Moscow is the capital of Russia. London is the capital city of the United Kingdom.

Output generated in 87.56 seconds (0.66 tokens/s, 58 tokens, context 58, seed 1168887889) Initial load
Output generated in 52.64 seconds (1.10 tokens/s, 58 tokens, context 58, seed 825034508)
Output generated in 50.83 seconds (1.14 tokens/s, 58 tokens, context 58, seed 1013750364)
Output generated in 52.58 seconds (1.10 tokens/s, 58 tokens, context 58, seed 1221966987)

**Results**
Only about a 26% increase. And 65B models still eventually boggle down for me. I can and do use them. Even 65B 5_1 I do use. However I can't say I can freely use them. Long discussions really grind to a crawl. GPU Acceleration still helps. It's not all about tokens per second. The extreme slow down and pauses with longer chats are still an issue I think with 65B with this configuration. Let me be clear long stories are fine. It's only when you start asking more prompts into an existing conversation will it become an issue on my computer. 

--------------------------------------------------------
I have one more result to share but it's not GGML comparison just want to show that GPTQ 13B is still extremely fast in comparison. 

**TheBloke_Wizard-Vicuna-13B-Uncensored-GPTQ** **(GPTQ-for-LLaMa)**
What is the capital of Canada. An the capital of USA. And China and Russia and the UK.

**Answer**
The capital of Canada is Ottawa. 
The capital of the United States is Washington D.C. 
The capital of China is Beijing. 
The capital of Russia is Moscow. 
The capital of the United Kingdom is London.

Output generated in 2.59 seconds (20.89 tokens/s, 54 tokens, context 61, seed 1367356941) Initial Run
Output generated in 2.51 seconds (21.55 tokens/s, 54 tokens, context 61, seed 1515313673) 
Output generated in 2.45 seconds (22.00 tokens/s, 54 tokens, context 61, seed 399424956)
Output generated in 2.45 seconds (22.07 tokens/s, 54 tokens, context 61, seed 1008728144)
Output generated in 2.54 seconds (21.28 tokens/s, 54 tokens, context 61, seed 541796200)

--------------------------------------------------------
**Conclusion** 
GGML GPU Acceleration is likely worth it. I am going to try to do some math. It's not perfect math and it's subject to hopefully "minor" error. But I want to demonstrate how long a paragraph takes to generate on a machine like mine. 

**Here are some helpful rules of thumb for understanding tokens in terms of lengths: 1 token ~= 4 chars in English.

    1 token ~= 4 chars in English

    1 token ~= ¾ words

    100 tokens ~= 75 words

Or 

    1-2 sentence ~= 30 tokens

    1 paragraph ~= 100 tokens

    1,500 words ~= 2048 tokens**
    
**GGML 65B 4_0** = about 70 seconds per paragraph give or take. Exact model and variance will slightly modify this. 
**GGML 30B 5_0** = about 36 seconds per paragraph etc. 
**GPTQ 13B** **(GPTQ-for-LLaMa)** = About 4.5 seconds per paragraph etc.

Yes I am aware that like 65B 5_0 or 5_1 will take longer etc. I do use models like that. These are only estimates based on a variety of factors. Settings, hardware, exact model and even what is running in the background will change these numbers. However it's interesting to at least plot some loose data and look it over. 

Software versions used
Visual Studios 2022 Community
Nvida Cuda Toolkit 12.1

**Comment by ossirytk** [comment]
This seems pretty in line with my experience too. I'm waiting to see how the couple of next performance upgrades for the core llama.cpp will turn out. There is a pr open for full Cuda gpu acceleration and a pr for new threading. There seem to be significant performance improvements. I 'm hoping that running 65B will be viable soon.

**Reply by berkut1** [reply]
Which one improvement? I have tested new k-quant methods, but didn't noticed so much improvements.

**Reply by ossirytk** [reply]
The Llama.cpp main repo has pr#1827 "CUDA full GPU acceleration, KV cache in VRAM" and pr#1632 "Fine tune MUL_MAT, new threading (spin+wait/notify), speedup q_f32 BLAS by splitting COMPUTE stage". These are still in pr stage in llama.cpp main repo.

**Reply by ossirytk** [reply]
The Cuda full gpu change also allows you to run ggml fully in gpu with single core. Performance seems pretty good.

**Reply by ossirytk** [reply]
The Cuda pr might also boost Cuda offloading since it adds things that were previously exclusive to cpu ram and are more effective when offloaded to gpu. Some values were previously only loaded to ram and you could not quite offload the whole model to gpu. Now everything can be offloaded and the big bottlenecks can be offloaded first. We'll see how this changes performance.

**Comment by viperwasp** [comment]
Update to include TheBloke_Wizard-Vicuna-13B-Uncensored-GPTQ GPTQ-for-LLaMa VS Auto GPTQ VS ExLlama (This does not change GGML test results. Only the GPTQ models.)

Prompts Various (I'm not actually posting the question/answers it's irreverent for this test as we are checking speeds.)

**Test 3** 
**TheBloke_Wizard-Vicuna-13B-Uncensored-GPTQ GPTQ-for-LLaMa**

**A.**  **GPTQ-for-LLaMa** GPU ONLY

Output generated in 2.59 seconds (20.89 tokens/s, 54 tokens, context 61, seed 1367356941) Initial Run
Output generated in 2.51 seconds (21.55 tokens/s, 54 tokens, context 61, seed 1515313673)
Output generated in 2.45 seconds (22.00 tokens/s, 54 tokens, context 61, seed 399424956)
Output generated in 2.45 seconds (22.07 tokens/s, 54 tokens, context 61, seed 1008728144)
Output generated in 2.54 seconds (21.28 tokens/s, 54 tokens, context 61, seed 541796200)

**B.** **Auto GPTQ**  GPU Only

Output generated in 2.11 seconds (21.78 tokens/s, 46 tokens, context 47, seed 1857161108) Initial Run
Output generated in 2.01 seconds (22.94 tokens/s, 46 tokens, context 47, seed 837633197)
Output generated in 4.12 seconds (25.02 tokens/s, 103 tokens, context 47, seed 1150961216)
Output generated in 4.06 seconds (25.35 tokens/s, 103 tokens, context 47, seed 1052977498)
Output generated in 4.19 seconds (24.59 tokens/s, 103 tokens, context 47, seed 43023413)
Output generated in 4.03 seconds (25.57 tokens/s, 103 tokens, context 47, seed 1597825514)

**C.** **ExLlama**  GPU Only 

Output generated in 3.80 seconds (52.67 tokens/s, 200 tokens, context 221, seed 814377025) Initial Run 
Output generated in 2.77 seconds (55.94 tokens/s, 155 tokens, context 45, seed 1303057945)
Output generated in 2.75 seconds (56.46 tokens/s, 155 tokens, context 45, seed 248859938)
Output generated in 3.71 seconds (53.88 tokens/s, 200 tokens, context 221, seed 1068447715)
Output generated in 3.64 seconds (54.91 tokens/s, 200 tokens, context 221, seed 1522449213)
Output generated in 3.69 seconds (54.16 tokens/s, 200 tokens, context 221, seed 1213784625)

**Results**
The results speak for themselves. However we don't yet know how model accuracy is effected. At least I don't. If anyone has any information please feel free to share it. Looking for like perplexity benchmark comparisons across all three backends etc. 

Updated final results seem to be...

**GGML 65B 4_0** = about 70 seconds per paragraph give or take. Exact model and variance will slightly modify this.
**GGML 30B 5_0** = about 36 seconds per paragraph etc.
**GPTQ 13B (GPTQ-for-LLaMa)** = About 4.5 seconds per paragraph etc.
**GPTQ 13B** **(ExLlama)** = About 1.84 seconds per paragraph etc.

**Reply by berkut1** [reply]
Can you check with context of 500 tokens (~20 sentences)?

**Reply by viperwasp** [reply]
Check what? The speed? In what format? GPTQ 13B (ExLlama)?

**Reply by viperwasp** [reply]
I don't know if this is what you wanted or not I can't get it do exactly 500 or even control it to well. 
Output generated in 9.36 seconds (54.27 tokens/s, 508 tokens, context 51, seed 2109024614)
Output generated in 8.83 seconds (54.22 tokens/s, 479 tokens, context 51, seed 90856253)
Output generated in 9.75 seconds (49.65 tokens/s, 484 tokens, context 51, seed 1563840292)
Output generated in 9.13 seconds (54.13 tokens/s, 494 tokens, context 51, seed 606418562)
Output generated in 8.44 seconds (49.86 tokens/s, 421 tokens, context 51, seed 497738302)

**Reply by berkut1** [reply]
The speed (GPTQ 13B (ExLlama)). You showed the data with context 51, not 500.

**Reply by viperwasp** [reply]
I think this will do it for you? 
Output generated in 1.99 seconds (52.79 tokens/s, 105 tokens, context 560, seed 187024135)
Output generated in 3.26 seconds (54.89 tokens/s, 179 tokens, context 530, seed 1677377389)
Output generated in 1.70 seconds (54.12 tokens/s, 92 tokens, context 498, seed 77441162)
Output generated in 1.68 seconds (54.60 tokens/s, 92 tokens, context 500, seed 2117720162)
Output generated in 1.63 seconds (56.40 tokens/s, 92 tokens, context 500, seed 1023878837)
Output generated in 1.63 seconds (56.48 tokens/s, 92 tokens, context 500, seed 1013198671)
Output generated in 1.64 seconds (55.93 tokens/s, 92 tokens, context 500, seed 203658733)
Sorry about that. 

I copied some legal data from a website and asked it to summarize not that I think the question or answer matter.

**Reply by berkut1** [reply]
Thanks, the results look great.

**Comment by grexzen** [comment]
What flags do you run or other settings? 

api
auto_launch
autogptq
cai_chat
compress_weight
deepspeed
flexgen
fused_mlp
gptq_for_llama
listen
model_menu
monkey_patch
multi_user
no_cache
no_stream
pin_weight
public_api
quant_attn
rwkv_cuda_on
sdp_attention
share
verbose
warmup_autotune
xformers
